docs/proposals/20210310-edge-device-management.md

```diff
@@ -0,0 +1,498 @@
+          description: "get current light color"
+          expectedValues: ["color"]
+        -
+          code: "503"
+          description: "service unavailable"
+          expectedValues: []
+      put:
+        path: "/api/v1/device/{deviceId}/changeColor"
+        responses:
+        -
+          code: "201"
+          description: "set the light color"
+        -
+          code: "503"
+          description: "service unavailable"
+          expectedValues: []
+```
+
+The `DeviceProfile` contains one `deviceResources`, i.e., `lightcolor`, which supports both "read" and "write" operations. The profile also defines a corresponding `coreCommand`, which contains the REST API of the `lightcolor`. After this `DeviceProfile` is applied to the cluster, the following core command will be created on the EdgeX Foundry,
+
+```json
+{
+  ...
+  "commands": [
+    {
+      "created": <created-timestamp>,
+      "modified": <modifed-timestamp>,
+      "id": "<command-id>",
+      "name": "lightcolor",
+      "get": {
+        "path": "/api/v1/device/{deviceId}/color",
+        "responses": [
+          {
+            "code": "200",
+            "description": "get current light color",
+            "expectedValues": [
+              "color"
+            ]
+          },
+          {
+            "code": "503",
+            "description": "service unavailable"
+          }
+        ],
+        "url": "http://edgex-core-command:48082/api/v1/device/<device-id>/command/<set-command-id>"
+      },
+      "put": {
+        "path": "/api/v1/device/{deviceId}/changeColor",
+        "responses": [
+          {
+            "code": "201",
+            "description": "set the light color"
+          },
+          {
+            "code": "503",
+            "description": "service unavailable"
+          }
+        ],
+        "url": "http://edgex-core-command:48082/api/v1/device/<device-id>/command/<put-command-id>"
+      }
+    }
+  ]
+}
+```
+
+Then we can create a `Device` associated to the `DeviceProfile` as following,
+
+```yaml
+apiVersion: device.openyurt.io/v1alpha1
+kind: Device
+metadata:
+  name: testsensor
+spec:
+  ...
+  profile: color-sensor
+  deviceProperties:
+    lightcolor:
+      name: lightcolor
+      desiredValue: green
+```
+
+Assume that the initial light color is blue, then after applying the `Device`, the device controller will try to fetch the corresponding get/put URLs from the core command service, update the `testsensor.Spec.DeviceProperties[lightcolor]` `testsensor.Status.DeviceProperties[lightcolor]`, and then set the light color to green, the updated `Device` will look like the following,
+```yaml
+apiVersion: device.openyurt.io/v1alpha1
+kind: Device
+metadata:
+  name: testsensor
+spec:
+  ...
+  profile: color-sensor
+  deviceProperties:
+    lightcolor:
+      name: lightcolor
+      setURL: http://edgex-core-command:48082/api/v1/device/<device-id>/command/<put-command-id>
+      desiredValue: green
+status:
```

Comment:
A little confused about the "status" section. do you mean the status is the device's current status data?

Reply:
Yes, the current observed status of the device.